File: src/main/java/studio/karllang/karl/lib/Function.java
package studio.karllang.karl.lib;

import java.util.ArrayList;
import studio.karllang.karl.parser.ast.expressions.Expression;

/**
 * Represents an abstract function within a library in a programming environment. Functions are
 * executable units of code that can take input expressions and produce output.
 */
public abstract class Function {
  private final String name; // The name of the function.
  private final Library library; // The library to which this function belongs.

  /**
   * Constructs a new Function object with the specified name and associated library.
   *
   * @param name The name of the function.
   * @param library The library to which the function belongs.
   */
  public Function(String name, Library library) {
    this.name = name;
    this.library = library;
  }

  /**
   * Executes the function with the given input expressions.
   *
   * @param expressions An ArrayList of expressions representing the input arguments to the
   *     function.
   */
  public abstract void eval(ArrayList<Expression> expressions);

  /**
   * Retrieves the name of the function.
   *
   * @return The name of the function.
   */
  public String getName() {
    return name;
  }

  /**
   * Retrieves the library to which this function belongs.
   *
   * @return The Library object representing the library to which the function belongs.
   */
  public Library getLibrary() {
    return library;
  }
}

File: src/main/kotlin/studio/karllang/karl/lib/Function.kt
package studio.karllang.karl.lib

import studio.karllang.karl.errors.runtime.RuntimeError
import studio.karllang.karl.lexer.TokenType
import studio.karllang.karl.parser.ast.expressions.Expression
import studio.karllang.karl.parser.ast.statements.BlockStatement
import studio.karllang.karl.parser.ast.values.Value

class Function(
    private val name: String,
    private val args: LinkedHashMap<String, TokenType>,
    private val type: TokenType,
    private val body: BlockStatement
) {
    fun eval(values: ArrayList<Expression>, line: Int, pos: Int): Value? {
        val arguments = HashMap<String, Value>()
        args.keys.zip(values).forEach { (k, v) -> arguments[k] = v.eval() }
        body.setArgs(arguments)
        body.eval()

        val result = body.result
        if (result != null) {
            if (type == TokenType.VOID) {
                throw RuntimeError("Function $name is void, but return a value", pos, line, printString())
            }
            if (Types.checkValueType(type, result.type) || (type == TokenType.STRING && result.type == TokenType.NULL)) {
                return result
            } else {
                throw RuntimeError("Incorrect return type for function $name: except ${type.name} but got type ${result.type.name}", pos, line, printString())
            }
        } else if (type != TokenType.VOID) {
            throw RuntimeError("Missing return statement in function: $name", pos, line, printString())
        }

        return null
    }

    fun getName(): String {
        return name
    }

    fun getArgs(): LinkedHashMap<String, TokenType> {
        return args
    }

    fun getType(): TokenType {
        return type
    }

    private fun printString(): String {
        val optionString = StringBuilder()
        for (key in args.keys) {
            optionString.append(if (optionString.isEmpty()) key else "$key, ")
        }
        return "$name($optionString);"
    }
}